src/librustdoc/html/markdown.rs

```diff
@@ -353,145 +212,300 @@ pub fn render(w: &mut fmt::Formatter,
+                    }
+                }
             } else {
-                let lang = (*lang).as_bytes();
-                let s = str::from_utf8(lang).unwrap();
-                LangString::parse(s)
-            };
-            if !block_info.rust { return }
-            let text = (*text).as_bytes();
-            let opaque = (*data).opaque as *mut hoedown_html_renderer_state;
-            let tests = &mut *((*opaque).opaque as *mut ::test::Collector);
-            let text = str::from_utf8(text).unwrap();
-            let lines = text.lines().map(|l| {
-                stripped_filtered_line(l).unwrap_or(l)
-            });
-            let text = lines.collect::<Vec<&str>>().join("\n");
-            let line = tests.get_line() + line;
-            let filename = tests.get_filename();
-            tests.add_test(text.to_owned(),
-                           block_info.should_panic, block_info.no_run,
-                           block_info.ignore, block_info.test_harness,
-                           block_info.compile_fail, block_info.error_codes,
-                           line, filename);
+                break
+            }
+        }
+        if content.is_empty() {
+            return
         }
+        buffer.push_str(&format!("<thead><tr>{}</tr></thead>", content.replace("td>", "th>")));
     }
 
-    extern fn header(_ob: *mut hoedown_buffer,
-                     text: *const hoedown_buffer,
-                     level: libc::c_int, data: *const hoedown_renderer_data,
-                     _: libc::size_t) {
-        unsafe {
-            let opaque = (*data).opaque as *mut hoedown_html_renderer_state;
-            let tests = &mut *((*opaque).opaque as *mut ::test::Collector);
-            if text.is_null() {
-                tests.register_header("", level as u32);
+    fn table(parser: &mut Parser, buffer: &mut String, toc_builder: &mut Option<TocBuilder>,
+             shorter: bool) {
+        let mut content = String::new();
+        let mut rows = String::new();
+        loop {
+            let event = parser.next();
+            if let Some(event) = event {
+                match event {
+                    Event::End(Tag::Table(_)) => break,
+                    Event::Start(Tag::TableHead) => {
+                        head(parser, &mut content, toc_builder, shorter);
+                    }
+                    Event::Start(Tag::TableRow) => {
+                        row(parser, &mut rows, toc_builder, shorter);
+                    }
+                    _ => {}
+                }
             } else {
-                let text = (*text).as_bytes();
-                let text = str::from_utf8(text).unwrap();
-                tests.register_header(text, level as u32);
+                break
             }
         }
+        buffer.push_str(&format!("<table>{}{}</table>",
+                                 content,
+                                 if shorter || rows.is_empty() {
```

Comment:
Is this the only place where `shorter` is actually used? The name `shorter` does not imply "no tbody tag" to me…